app/javascript/mastodon/features/ui/components/doodle_modal.js

```diff
@@ -3,55 +3,319 @@ import PropTypes from 'prop-types';
+for (let row = 0; row < 7; row++) {
+  for (let col = 0; col < 11; col++) {
+    palReordered.push(palette[col * 7 + row]);
+  }
+  palReordered.push(null); // null indicates a <br />
+}
+
+// Utility for converting base64 image to binary for upload
+// https://stackoverflow.com/questions/35940290/how-to-convert-base64-string-to-javascript-file-object-like-as-from-file-input-f
+function dataURLtoFile(dataurl, filename) {
+  let arr = dataurl.split(','), mime = arr[0].match(/:(.*?);/)[1],
+    bstr = atob(arr[1]), n = bstr.length, u8arr = new Uint8Array(n);
+  while(n--){
+    u8arr[n] = bstr.charCodeAt(n);
+  }
+  return new File([u8arr], filename, { type: mime });
+}
+
+
+const mapStateToProps = state => ({
+  options: state.getIn(['compose', 'doodle']),
+});
+
+const mapDispatchToProps = dispatch => ({
+  setOpt: (opts) => dispatch(doodleSet(opts)),
+  submit: (file) => dispatch(uploadCompose([file])),
+});
+
+@connect(mapStateToProps, mapDispatchToProps)
+export default class DoodleModal extends ImmutablePureComponent {
 
   static propTypes = {
-    onDoodleSubmit: PropTypes.func.isRequired, // gets the base64 as argument
+    options: ImmutablePropTypes.map,
     onClose: PropTypes.func.isRequired,
+    setOpt: PropTypes.func.isRequired,
+    submit: PropTypes.func.isRequired,
   };
 
+  //region Option getters/setters
+
+  get fg () {
+    return this.props.options.get('fg');
+  }
+
+  set fg (value) {
+    this.props.setOpt({ fg: value });
+  }
+
+  get bg () {
+    return this.props.options.get('bg');
+  }
+
+  set bg (value) {
+    this.props.setOpt({ bg: value });
+  }
+
+  get mode () {
+    return this.props.options.get('mode');
+  }
+
+  set mode (value) {
+    this.props.setOpt({ mode: value });
+  }
+
+  get weight () {
+    return this.props.options.get('weight');
+  }
+
+  set weight (value) {
+    this.props.setOpt({ weight: value });
+  }
+
+  get opacity () {
+    return this.props.options.get('opacity');
+  }
+
+  set opacity (value) {
+    this.props.setOpt({ opacity: value });
+  }
+
+  get adaptiveStroke () {
+    return this.props.options.get('adaptiveStroke');
+  }
+
+  set adaptiveStroke (value) {
+    this.props.setOpt({ adaptiveStroke: value });
+  }
+
+  get smoothing () {
+    return this.props.options.get('smoothing');
+  }
+
+  set smoothing (value) {
+    this.props.setOpt({ smoothing: value });
+  }
+
+  //endregion
+
   handleKeyUp = (e) => {
     if (e.key === 'Delete' || e.key === 'Backspace') {
+      e.preventDefault();
       this.clearScreen();
     }
-  }
 
-  clearScreen () {
-    this.sketcher.context.fillStyle = 'white';
-    this.sketcher.context.fillRect(0, 0, this.canvas.width, this.canvas.height);
-  }
+    if (e.key === 'z' && e.ctrlKey) {
+      e.preventDefault();
+      this.undo();
+    }
+  };
 
   componentDidMount () {
     window.addEventListener('keyup', this.handleKeyUp, false);
+  };
+
+  componentWillUnmount () {
+    window.removeEventListener('keyup', this.handleKeyUp);
```

Comment:
I don't *think* it will cause any problems as-is but you should probably add a `, false` to this so that it exactly matches the call in `componentDidMount()`

Reply:
added